dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
+        }
+
+        AddExecutionStatsToContext(stepsTaken, context, this.Config.MaxIterations);
+        context.Variables.Update($"Result not found, review _stepsTaken to see what happened.\n{JsonSerializer.Serialize(stepsTaken)}");
+
+        return context;
+    }
+
+    #region setup helpers
+    // Function to load PromptConfig from embedded resource
+    private static PromptTemplateConfig LoadPromptConfigFromResource()
+    {
+        string promptConfigString = EmbeddedResource.Read("Skills.StepwiseStep.config.json");
+        return !string.IsNullOrEmpty(promptConfigString) ? PromptTemplateConfig.FromJson(promptConfigString) : new PromptTemplateConfig();
+    }
+
+    private async Task<ChatHistory> InitializeChatHistoryAsync(ChatHistory chatHistory, IAIService aiService, SKContext context)
+    {
+        // this.CreateChatHistory(this._kernel, out var aiService);
+        var chatCompletion = aiService as IChatCompletion;
+        var textCompletion = aiService as ITextCompletion;
+
+        string userManual = await this.GetUserManualAsync(context).ConfigureAwait(false);
+        string userQuestion = await this.GetUserQuestionAsync(context).ConfigureAwait(false);
+
+#pragma warning disable CA2016 // Forward the 'CancellationToken' parameter to methods that take one -- the method is obsolete and to be removed.
+        var systemContext = this._kernel.CreateNewContext();
+#pragma warning restore CA2016
+
+        systemContext.Variables.Set("suffix", this.Config.Suffix);
+
+        if (chatCompletion is null)
+        {
+            systemContext.Variables.Set("functionDescriptions", userManual);
+        }
+        string systemMessage = await this.GetSystemMessage(systemContext).ConfigureAwait(false);
+
+        chatHistory.AddSystemMessage(systemMessage);
+        if (chatCompletion is not null)
+        {
+            chatHistory.AddUserMessage(userManual);
+        }
+        chatHistory.AddUserMessage(userQuestion);
+
+        return chatHistory;
+    }
+
+    private ChatHistory CreateChatHistory(IKernel kernel, out IAIService aiService)
+    {
+        ChatHistory chatHistory;
+        if (TryGetChatCompletion(this._kernel, out var chatCompletion))
+        {
+            chatHistory = chatCompletion.CreateNewChat();
+            aiService = chatCompletion;
         }
         else
         {
-            context.Variables.Update("Question not found.");
+            var textCompletion = this._kernel.GetService<ITextCompletion>();
+            aiService = textCompletion;
+            chatHistory = new ChatHistory();
         }
 
-        return context;
+        return chatHistory;
+    }
+
+    private static bool TryGetChatCompletion(IKernel kernel, [NotNullWhen(true)] out IChatCompletion? chatCompletion)
+    {
+        try
+        {
+            // Client used to request answers to chat completion models
+            chatCompletion = kernel.GetService<IChatCompletion>();
```

Comment:
>GetService [](http://example.com/codeflow?start=36&length=10)

Using exception handling in the normal flow is fairly expensive. Can we add a TryGetService to the kernel API?

Reply:
Opened #2635 to track and will leave note here for now.